File: applications/settings/about/about.c
#include <furi.h>
#include <dialogs/dialogs.h>
#include <gui/gui.h>
#include <gui/view_dispatcher.h>
#include <gui/modules/empty_screen.h>
#include <gui/elements.h>
#include <assets_icons.h>
#include <furi_hal_version.h>
#include <furi_hal_region.h>
#include <furi_hal_bt.h>
#include <power/power_service/power.h>

int screen_index;

#define LOW_CHARGE_THRESHOLD 10
#define HIGH_DRAIN_CURRENT_THRESHOLD 100

typedef DialogMessageButton (*AboutDialogScreen)(DialogsApp* dialogs, DialogMessage* message);

static DialogMessageButton product_screen(DialogsApp* dialogs, DialogMessage* message) {
    DialogMessageButton result;

    FuriString* screen_header = furi_string_alloc_printf(
        "Product: %s\n"
        "Model: %s",
        furi_hal_version_get_model_name(),
        furi_hal_version_get_model_code());

    FuriString* screen_text = furi_string_alloc_printf(
        "FCC ID: %s\n"
        "IC: %s",
        furi_hal_version_get_fcc_id(),
        furi_hal_version_get_ic_id());

    dialog_message_set_header(
        message, furi_string_get_cstr(screen_header), 0, 0, AlignLeft, AlignTop);
    dialog_message_set_text(
        message, furi_string_get_cstr(screen_text), 0, 26, AlignLeft, AlignTop);
    result = dialog_message_show(dialogs, message);
    dialog_message_set_header(message, NULL, 0, 0, AlignLeft, AlignTop);
    dialog_message_set_text(message, NULL, 0, 0, AlignLeft, AlignTop);

    furi_string_free(screen_header);
    furi_string_free(screen_text);

    return result;
}

static DialogMessageButton address_screen(DialogsApp* dialogs, DialogMessage* message) {
    DialogMessageButton result;

    const char* screen_text = "Flipper Devices Inc\n"
                              "Suite B #551, 2803\n"
                              "Philadelphia Pike, Claymont\n"
                              "DE, USA 19703\n";

    dialog_message_set_text(message, screen_text, 0, 0, AlignLeft, AlignTop);
    result = dialog_message_show(dialogs, message);
    dialog_message_set_text(message, NULL, 0, 0, AlignLeft, AlignTop);

    return result;
}

static DialogMessageButton compliance_screen(DialogsApp* dialogs, DialogMessage* message) {
    DialogMessageButton result;

    const char* screen_text = "For all compliance\n"
                              "certificates please visit:\n"
                              "www.flipp.dev/compliance";

    dialog_message_set_text(message, screen_text, 0, 0, AlignLeft, AlignTop);
    result = dialog_message_show(dialogs, message);
    dialog_message_set_text(message, NULL, 0, 0, AlignLeft, AlignTop);

    return result;
}

static DialogMessageButton icon1_screen(DialogsApp* dialogs, DialogMessage* message) {
    DialogMessageButton result;

    dialog_message_set_icon(message, &I_Certification1_103x56, 13, 0);
    result = dialog_message_show(dialogs, message);
    dialog_message_set_icon(message, NULL, 0, 0);

    return result;
}

static DialogMessageButton icon2_screen(DialogsApp* dialogs, DialogMessage* message) {
    DialogMessageButton result;

    dialog_message_set_icon(message, &I_Certification2_98x33, 15, 10);
    result = dialog_message_show(dialogs, message);
    dialog_message_set_icon(message, NULL, 0, 0);

    return result;
}

static DialogMessageButton hw_version_screen(DialogsApp* dialogs, DialogMessage* message) {
    DialogMessageButton result;
    FuriString* buffer;
    buffer = furi_string_alloc();
    const char* my_name = furi_hal_version_get_name_ptr();

    furi_string_cat_printf(
        buffer,
        "%d.F%dB%dC%d %s:%s %s\n",
        furi_hal_version_get_hw_version(),
        furi_hal_version_get_hw_target(),
        furi_hal_version_get_hw_body(),
        furi_hal_version_get_hw_connect(),
        furi_hal_version_get_hw_region_name_otp(),
        furi_hal_region_get_name(),
        my_name ? my_name : "Unknown");

    furi_string_cat_printf(buffer, "Serial Number:\n");
    const uint8_t* uid = furi_hal_version_uid();
    for(size_t i = 0; i < furi_hal_version_uid_size(); i++) {
        furi_string_cat_printf(buffer, "%02X", uid[i]);
    }

    dialog_message_set_header(message, "HW Version Info:", 0, 0, AlignLeft, AlignTop);
    dialog_message_set_text(message, furi_string_get_cstr(buffer), 0, 13, AlignLeft, AlignTop);
    result = dialog_message_show(dialogs, message);
    dialog_message_set_text(message, NULL, 0, 0, AlignLeft, AlignTop);
    dialog_message_set_header(message, NULL, 0, 0, AlignLeft, AlignTop);
    furi_string_free(buffer);

    return result;
}

static DialogMessageButton fw_version_screen(DialogsApp* dialogs, DialogMessage* message) {
    DialogMessageButton result;
    FuriString* buffer;
    buffer = furi_string_alloc();
    const Version* ver = furi_hal_version_get_firmware_version();
    const BleGlueC2Info* c2_ver = NULL;
#ifdef SRV_BT
    c2_ver = ble_glue_get_c2_info();
#endif

    if(!ver) { //-V1051
        furi_string_cat_printf(buffer, "No info\n");
    } else {
        furi_string_cat_printf(
            buffer,
            "%s [%s]\n%s%s [%s] %s\n[%d] %s",
            version_get_version(ver),
            version_get_builddate(ver),
            version_get_dirty_flag(ver) ? "[!] " : "",
            version_get_githash(ver),
            version_get_gitbranchnum(ver),
            c2_ver ? c2_ver->StackTypeString : "<none>",
            version_get_target(ver),
            version_get_gitbranch(ver));
    }

    dialog_message_set_header(message, "FW Version Info:", 0, 0, AlignLeft, AlignTop);
    dialog_message_set_text(message, furi_string_get_cstr(buffer), 0, 13, AlignLeft, AlignTop);
    result = dialog_message_show(dialogs, message);
    dialog_message_set_text(message, NULL, 0, 0, AlignLeft, AlignTop);
    dialog_message_set_header(message, NULL, 0, 0, AlignLeft, AlignTop);
    furi_string_free(buffer);

    return result;
}

static void draw_stat(Canvas* canvas, int x, int y, const Icon* icon, char* val) {
    canvas_draw_frame(canvas, x - 7, y + 7, 30, 13);
    canvas_draw_icon(canvas, x, y, icon);
    canvas_set_color(canvas, ColorWhite);
    canvas_draw_box(canvas, x - 4, y + 16, 24, 6);
    canvas_set_color(canvas, ColorBlack);
    canvas_draw_str_aligned(canvas, x + 8, y + 22, AlignCenter, AlignBottom, val);
};

static void draw_battery(Canvas* canvas, PowerInfo* info, int x, int y) {
    char header[20] = {};
    char value[20] = {};

    int32_t drain_current = info->current_gauge * (-1000);
    uint32_t charge_current = info->current_gauge * 1000;

    // Draw battery
    canvas_draw_icon(canvas, x, y, &I_BatteryBody_52x28);
    if(charge_current > 0) {
        canvas_draw_icon(canvas, x + 16, y + 7, &I_FaceCharging_29x14);
    } else if(drain_current > HIGH_DRAIN_CURRENT_THRESHOLD) {
        canvas_draw_icon(canvas, x + 16, y + 7, &I_FaceConfused_29x14);
    } else if(info->charge < LOW_CHARGE_THRESHOLD) {
        canvas_draw_icon(canvas, x + 16, y + 7, &I_FaceNopower_29x14);
    } else {
        canvas_draw_icon(canvas, x + 16, y + 7, &I_FaceNormal_29x14);
    }

    // Draw bubble
    elements_bubble(canvas, x + 53, y + 0, 71, 28);

    // Set text
    if(charge_current > 0) {
        snprintf(header, sizeof(header), "%s", "Charging at");
        snprintf(
            value,
            sizeof(value),
            "%lu.%luV   %lumA",
            (uint32_t)(info->voltage_vbus),
            (uint32_t)(info->voltage_vbus * 10) % 10,
            charge_current);
    } else if(drain_current > 0) {
        snprintf(header, sizeof(header), "%s", "Consumption is");
        snprintf(
            value,
            sizeof(value),
            "%ld %s",
            drain_current,
            drain_current > HIGH_DRAIN_CURRENT_THRESHOLD ? "mA!" : "mA");
    } else if(drain_current != 0) {
        snprintf(header, 20, "...");
    } else if(info->voltage_battery_charging < 4.2) {
        // Non-default battery charging limit, mention it
        snprintf(header, sizeof(header), "Limited to");
        snprintf(
            value,
            sizeof(value),
            "%lu.%luV",
            (uint32_t)(info->voltage_battery_charging),
            (uint32_t)(info->voltage_battery_charging * 10) % 10);
    } else {
        snprintf(header, sizeof(header), "Charged!");
    }

    if(!strcmp(value, "")) {
        canvas_draw_str_aligned(canvas, x + 92, y + 14, AlignCenter, AlignCenter, header);
    } else if(!strcmp(header, "")) {
        canvas_draw_str_aligned(canvas, x + 92, y + 14, AlignCenter, AlignCenter, value);
    } else {
        canvas_draw_str_aligned(canvas, x + 92, y + 9, AlignCenter, AlignCenter, header);
        canvas_draw_str_aligned(canvas, x + 92, y + 19, AlignCenter, AlignCenter, value);
    }
};

static void battery_info_draw_callback(Canvas* canvas, void* context) {
    furi_assert(context);
    PowerInfo* info = context;

    canvas_clear(canvas);
    canvas_set_color(canvas, ColorBlack);
    draw_battery(canvas, info, 0, 0);

    char batt_level[10];
    char temperature[10];
    char voltage[10];
    char health[10];

    snprintf(batt_level, sizeof(batt_level), "%lu%%", (uint32_t)info->charge);
    snprintf(temperature, sizeof(temperature), "%lu C", (uint32_t)info->temperature_gauge);
    snprintf(
        voltage,
        sizeof(voltage),
        "%lu.%01lu V",
        (uint32_t)info->voltage_gauge,
        (uint32_t)(info->voltage_gauge * 10) % 10UL);
    snprintf(health, sizeof(health), "%d%%", info->health);

    draw_stat(canvas, 8, 28, &I_Battery_16x16, batt_level);
    draw_stat(canvas, 40, 28, &I_Temperature_16x16, temperature);
    draw_stat(canvas, 72, 28, &I_Voltage_16x16, voltage);
    draw_stat(canvas, 104, 28, &I_Health_16x16, health);

    elements_button_left(canvas, "Back");
    elements_button_right(canvas, "Next");
}

static bool battery_info_input_callback(InputEvent* event, void* context) {
    FuriSemaphore* semaphore = context;

    bool consumed = false;
    if(event->type == InputTypeShort) {
        if(event->key == InputKeyLeft) {
            screen_index--;
            consumed = true;
        } else if(event->key == InputKeyRight) {
            screen_index++;
            consumed = true;
        } else if(event->key == InputKeyBack) {
            screen_index = -2;
            consumed = true;
        }
    }

    if(consumed) {
        furi_semaphore_release(semaphore);
    }
    return consumed;
}

const AboutDialogScreen about_screens[] = {
    product_screen,
    compliance_screen,
    address_screen,
    icon1_screen,
    icon2_screen,
    hw_version_screen,
    fw_version_screen};

const int about_screens_count = sizeof(about_screens) / sizeof(AboutDialogScreen);

int32_t about_settings_app(void* p) {
    bool battery_info = false;
    if(p && strlen(p) && !strcmp(p, "batt")) {
        battery_info = true;
    }

    Gui* gui = furi_record_open(RECORD_GUI);
    ViewDispatcher* view_dispatcher = view_dispatcher_alloc();
    const uint32_t battery_info_index = 0;
    const uint32_t empty_screen_index = 1;

    FuriSemaphore* semaphore = furi_semaphore_alloc(1, 0);
    Power* power = furi_record_open(RECORD_POWER);
    View* battery_view = view_alloc();
    view_set_context(battery_view, semaphore);
    view_set_input_callback(battery_view, battery_info_input_callback);
    view_allocate_model(battery_view, ViewModelTypeLocking, sizeof(PowerInfo));
    view_set_draw_callback(battery_view, battery_info_draw_callback);

    DialogsApp* dialogs = furi_record_open(RECORD_DIALOGS);
    DialogMessage* message = dialog_message_alloc();
    EmptyScreen* empty_screen = empty_screen_alloc();
    DialogMessageButton screen_result;

    // draw empty screen to prevent menu flickering
    view_dispatcher_add_view(view_dispatcher, battery_info_index, battery_view);
    view_dispatcher_add_view(
        view_dispatcher, empty_screen_index, empty_screen_get_view(empty_screen));
    view_dispatcher_attach_to_gui(view_dispatcher, gui, ViewDispatcherTypeFullscreen);

    screen_index = -1 + !battery_info;
    while(screen_index > -2) {
        if(screen_index == -1) {
            if(!battery_info) {
                break;
            }
            with_view_model(
                battery_view, PowerInfo * model, { power_get_info(power, model); }, true);
            view_dispatcher_switch_to_view(view_dispatcher, battery_info_index);
            furi_semaphore_acquire(semaphore, 2000);
        } else {
            view_dispatcher_switch_to_view(view_dispatcher, empty_screen_index);
            if(screen_index >= about_screens_count - 1) {
                dialog_message_set_buttons(message, "Back", NULL, NULL);
            } else {
                dialog_message_set_buttons(message, "Back", NULL, "Next");
            }
            screen_result = about_screens[screen_index](dialogs, message);
            if(screen_result == DialogMessageButtonLeft) {
                screen_index--;
            } else if(screen_result == DialogMessageButtonRight) {
                screen_index++;
            } else if(screen_result == DialogMessageButtonBack) {
                screen_index = -2;
            }
        }
    }

    dialog_message_free(message);
    furi_record_close(RECORD_DIALOGS);

    furi_record_close(RECORD_POWER);
    furi_semaphore_free(semaphore);
    view_dispatcher_remove_view(view_dispatcher, empty_screen_index);
    view_dispatcher_remove_view(view_dispatcher, battery_info_index);
    view_free(battery_view);
    view_dispatcher_free(view_dispatcher);
    empty_screen_free(empty_screen);
    furi_record_close(RECORD_GUI);

    return 0;
}

File: applications/system/elk_mjs/assets/about.js

let RECORD = ({
    open: function (api) {
        let ret = Object.create(this.prototype);
        ret.name = api.name;
        ret.ptr = ret.api.open(ret.name);
        return ret;
    },
    prototype: {
        api: {
            open: ffi("void* furi_record_open(char *)"),
            close: ffi("void furi_record_close(char *)"),
        },
        close: function () {
            this.api.close(this.name);
        }
    }
});

let GUI = ({
    name: "gui",
    align: {
        left: 0,
        right: 1,
        top: 2,
        bottom: 3,
        center: 4,
    },
});

let DIALOGS = ({
    name: "dialogs",
    show: function (record, message) {
        return ffi("int dialog_message_show(void*, void*)")(record.ptr, message.ptr);
    },
});

let DIALOG_MESSAGE = ({
    alloc: function () {
        let ret = Object.create(this.prototype);
        ret.ptr = ret.api.alloc();
        return ret;
    },
    button_type: {
        back: 0,
        left: 1,
        center: 2,
        right: 3,
    },
    prototype: {
        api: {
            alloc: ffi("void* dialog_message_alloc()"),
            free: ffi("void dialog_message_free(void*)"),
            set_header: ffi("void dialog_message_set_header(void*, char*, int, int, int, int)"),
            set_text: ffi("void dialog_message_set_text(void*, char*, int, int, int, int)"),
            set_buttons: ffi("void dialog_message_set_buttons(void*, char*, char*, char*)"),
            set_icon: ffi("void dialog_message_set_icon(void*, void*, int, int)"),
        },
        free: function () {
            this.api.free(this.ptr);
        },
        set_header: function (text, x, y, align_x, align_y) {
            this.api.set_header(this.ptr, text, x, y, align_x, align_y);
        },
        set_text: function (text, x, y, align_x, align_y) {
            this.api.set_text(this.ptr, text, x, y, align_x, align_y);
        },
        set_buttons: function (text_left, text_center, text_right) {
            this.api.set_buttons(this.ptr, text_left, text_center, text_right);
        },
        set_icon: function (icon, x, y) {
            this.api.set_icon(this.ptr, icon, x, y);
        },
        set_icon_from_fw: function (icon_name, x, y) {
            this.api.set_icon(this.ptr, ffi_address(icon_name), x, y);
        },
    }
});

let VIEW_DISPATCHER = ({
    alloc: function (record, type) {
        let ret = Object.create(this.prototype);
        ret.ptr = ret.api.alloc();
        ret.api.attach_to_gui(ret.ptr, record.ptr, type);
        return ret;
    },
    type: {
        desktop: 0,
        window: 1,
        fullscreen: 2,
    },
    prototype: {
        api: {
            alloc: ffi("void* view_dispatcher_alloc()"),
            free: ffi("void view_dispatcher_free(void*)"),
            add_view: ffi("void view_dispatcher_add_view(void*, int, void*)"),
            remove_view: ffi("void view_dispatcher_remove_view(void*, int)"),
            attach_to_gui: ffi("void view_dispatcher_attach_to_gui(void*, void*, int)"),
            switch_to_view: ffi("void view_dispatcher_switch_to_view(void*, int)"),
        },
        free: function () {
            this.api.free(this.ptr);
        },
        add_view: function (view) {
            this.api.add_view(this.ptr, view.index, view.get_view());
        },
        remove_view: function (view) {
            this.api.remove_view(this.ptr, view.index);
        },
        switch_to_view: function (view) {
            this.api.switch_to_view(this.ptr, view.index);
        },
    }
});

let EMPTY_SCREEN = ({
    alloc: function (index) {
        let ret = Object.create(this.prototype);
        ret.ptr = ret.api.alloc();
        ret.index = index;
        return ret;
    },
    prototype: {
        api: {
            alloc: ffi("void* empty_screen_alloc()"),
            free: ffi("void empty_screen_free(void*)"),
            get_view: ffi("void* empty_screen_get_view(void*)"),
        },
        free: function () {
            this.api.free(this.ptr);
        },
        get_view: function () {
            return this.api.get_view(this.ptr);
        },
    }
});

let STD = ({
    malloc: ffi('void *malloc(int)'),
    free: ffi('void free(void*)'),
});

let C_STR = ({
    create: function (str) {
        let len = str.length;
        let ret = Object.create(this.prototype);
        ret.mem = STD.malloc(len + 1);
        ret.text = mkstr(ret.mem, len);
        for (let i = 0; i < len; i++) {
            ret.mem[i] = str.at(i);
        }
        return ret;
    },
    prototype: {
        free: function () {
            STD.free(this.mem);
        },
        set: function (str) {
            let len = str.length;
            STD.free(this.mem);
            this.mem = STD.malloc(len + 1);
            this.text = mkstr(this.mem, len);
            for (let i = 0; i < len; i++) {
                this.mem[i] = str.at(i);
            }
        }
    }
});

let FURI_HAL_VERSION = ({
    get_name: ffi("char *furi_hal_version_get_name_ptr()"),
    get_hw_version: ffi("int furi_hal_version_get_hw_version()"),
    get_hw_target: ffi("int furi_hal_version_get_hw_target()"),
    get_hw_body: ffi("int furi_hal_version_get_hw_body()"),
    get_hw_connect: ffi("int furi_hal_version_get_hw_connect()"),
    get_hw_region_name: ffi("char* furi_hal_version_get_hw_region_name()"),
    get_region_name: ffi("char* furi_hal_region_get_name()"),
    version_uid: ffi("char* furi_hal_version_uid()"),
    version_uid_size: ffi("int furi_hal_version_uid_size()"),

    get_firmware_version: ffi("void* furi_hal_version_get_firmware_version()"),
    get_version: ffi("char* version_get_version(void*)"),
    get_builddate: ffi("char* version_get_builddate(void*)"),
    get_dirty_flag: ffi("int version_get_dirty_flag(void*)"),
    get_githash: ffi("char* version_get_githash(void*)"),
    get_gitbranchnum: ffi("char* version_get_gitbranchnum(void*)"),
    get_target: ffi("int version_get_target(void*)"),
    get_gitbranch: ffi("char* version_get_gitbranch(void*)"),
});

// init c-string
let text_back = C_STR.create("Back");
let text_next = C_STR.create("Next");
let text_header = C_STR.create("");
let text_body = C_STR.create("");

// open records
let dialogs = RECORD.open(DIALOGS);
let gui = RECORD.open(GUI);

// create objects
let message = DIALOG_MESSAGE.alloc();
let view_dispatcher = VIEW_DISPATCHER.alloc(gui, VIEW_DISPATCHER.type.fullscreen);
let empty_screen = EMPTY_SCREEN.alloc(0);

// configure objects
view_dispatcher.add_view(empty_screen);
view_dispatcher.switch_to_view(empty_screen);

// init result
let screen_index = 0;
let screen_result = DIALOG_MESSAGE.button_type.right;

let screens = [
    function () {
        text_header.set("Product: Flipper Zero\nModel: FZ.1\n");
        text_body.set("FCC ID: 2A2V6-FZ\nIC: 27624-FZ");
        message.set_header(text_header.text, 0, 0, GUI.align.left, GUI.align.top);
        message.set_text(text_body.text, 0, 26, GUI.align.left, GUI.align.top);
        let result = DIALOGS.show(dialogs, message);
        message.set_header(null, 0, 0, GUI.align.left, GUI.align.top);
        message.set_text(null, 0, 0, GUI.align.left, GUI.align.top);
        return result;
    },
    function () {
        text_body.set("Flipper Devices Inc\nSuite B #551, 2803\nPhiladelphia Pike, Claymont\nDE, USA 19703\n");
        message.set_text(text_body.text, 0, 0, GUI.align.left, GUI.align.top);
        let result = DIALOGS.show(dialogs, message);
        message.set_text(null, 0, 0, GUI.align.left, GUI.align.top);
        return result;
    },
    function () {
        text_body.set("For all compliance\ncertificates please visit:\nwww.flipp.dev/compliance");
        message.set_text(text_body.text, 0, 0, GUI.align.left, GUI.align.top);
        let result = DIALOGS.show(dialogs, message);
        message.set_text(null, 0, 0, GUI.align.left, GUI.align.top);
        return result;
    },
    function () {
        message.set_icon_from_fw("I_Certification1_103x56", 13, 0);
        let result = DIALOGS.show(dialogs, message);
        message.set_icon(null, 0, 0);
        return result;
    },
    function () {
        message.set_icon_from_fw("I_Certification2_46x33", 15, 10);
        let result = DIALOGS.show(dialogs, message);
        message.set_icon(null, 0, 0);
        return result;
    },
    function () {
        let my_name = FURI_HAL_VERSION.get_name();
        if (my_name === null) {
            my_name = "Unknown";
        }
        let str = to_string(FURI_HAL_VERSION.get_hw_version()) + ".";
        str += "F" + to_string(FURI_HAL_VERSION.get_hw_target());
        str += "B" + to_string(FURI_HAL_VERSION.get_hw_body());
        str += "C" + to_string(FURI_HAL_VERSION.get_hw_connect());
        str += " " + FURI_HAL_VERSION.get_hw_region_name() + ":" + FURI_HAL_VERSION.get_region_name();
        str += " " + my_name + "\n";
        str += "Serial Number:\n";


        let uid = FURI_HAL_VERSION.version_uid();
        let uid_size = FURI_HAL_VERSION.version_uid_size();
        for (let i = 0; i < uid_size; i++) {
            str += to_hex_string(uid.at(i));
        }

        text_header.set("HW Version Info:");
        text_body.set(str);
        message.set_header(text_header.text, 0, 0, GUI.align.left, GUI.align.top);
        message.set_text(text_body.text, 0, 13, GUI.align.left, GUI.align.top);
        let result = DIALOGS.show(dialogs, message);
        message.set_header(null, 0, 0, GUI.align.left, GUI.align.top);
        message.set_text(null, 0, 0, GUI.align.left, GUI.align.top);
        return result;
    },
    function () {
        let ver = FURI_HAL_VERSION.get_firmware_version();
        let str = "";

        if (ver === null) {
            str = "No info\n";
        } else {
            str = FURI_HAL_VERSION.get_version(ver) + " [";
            str += FURI_HAL_VERSION.get_builddate(ver) + "]\n";
            if (FURI_HAL_VERSION.get_dirty_flag(ver) > 0) {
                str += "[!]";
            }
            str += FURI_HAL_VERSION.get_githash(ver) + " [";
            str += FURI_HAL_VERSION.get_gitbranchnum(ver) + "]\n";

            str += "[" + to_string(FURI_HAL_VERSION.get_target(ver)) + "] ";
            str += FURI_HAL_VERSION.get_gitbranch(ver);
        }

        text_header.set("FW Version Info:");
        text_body.set(str);
        message.set_header(text_header.text, 0, 0, GUI.align.left, GUI.align.top);
        message.set_text(text_body.text, 0, 13, GUI.align.left, GUI.align.top);
        let result = DIALOGS.show(dialogs, message);
        message.set_header(null, 0, 0, GUI.align.left, GUI.align.top);
        message.set_text(null, 0, 0, GUI.align.left, GUI.align.top);
        return result;
    },
];

// app body
while (screen_result !== DIALOG_MESSAGE.button_type.back) {
    if (screen_index >= screens.length - 1) {
        message.set_buttons(text_back.text, null, null);
    } else {
        message.set_buttons(text_back.text, null, text_next.text);
    }

    screen_result = screens[screen_index]();

    if (screen_result === DIALOG_MESSAGE.button_type.left) {
        if (screen_index <= 0) {
            break;
        } else {
            screen_index--;
        }
    } else if (screen_result === DIALOG_MESSAGE.button_type.right) {
        if (screen_index < screens.length) {
            screen_index++;
        }
    } else if (screen_result === DIALOG_MESSAGE.button_type.back) {
        break;
    }
}

// free c-string
message.free();

// free objects
view_dispatcher.remove_view(empty_screen);
view_dispatcher.free();
empty_screen.free();

text_back.free();
text_next.free();
text_header.free();
text_body.free();

// close records
dialogs.close();
gui.close();